tests/neg/i5397.scala

```diff
@@ -16,8 +16,10 @@ object Test {
         rec3 // error: not in tail position
     })
 
-  @tailrec def rec4: Unit = {
-    def local = rec4 // error: not in tail position
+  // This is technically not breaching tail recursion as rec4 does not call itself, local does
+  // This instead fails due to having no tail recursion at all
+  @tailrec def rec4: Unit = { // error: no recursive calls
+    def local = rec4
```

Comment:
The more I think about this, the more counterintuitive it seems.
I'm fairly confident that we will get people complaining about cases like this, even if technically correct.

Do we need to consider having an option on the annotation to allow or deny indirect calls?
This could also allow handeling of non-calls such as:
```Scala
  def bar(f: Int => Long): Long =
    ???

  @tailrec
  def foo(i: Int): Long =
    if (???)
      bar(foo)
    else
      foo(i + 1)
```

Reply:
It is of course possible to craft examples that will be confusing. We should focus on what the usual scenario is. The usual scenario is that I'm writing a tail-recursive method, and I want the compiler to make sure I did not mess it up.

Indirect calls are usually a case of "I messed it up". For example, using `Option.fold` instead of pattern-matching prevents tail-recursive call elimination:
```scala
def foo(x: Option[Foo]): Unit =
  x.fold {
    0
  } { someFoo =>
    foo(None) // oops, this is not tail-recursive!
  }
```
I don't think complicating `@tailrec` with options will help that usual scenario. It will only make it more difficult to teach what `@tailrec` does.

Reply:
Ok, read the prior debates on this issue.
While I don't fully agree, its clear that this is the group intent, and very happy that we can now simply pick up this warning in CI tools.